Consider adding retry behavior in transient network failures

In a recent failed test run I see this:
```console
[2019/03/20 16:10:53][INFO] $ pushd "/home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/src/benchmarks/micro"
[2019/03/20 16:10:53][INFO] $ dotnet restore /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/src/benchmarks/micro/MicroBenchmarks.csproj --packages /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/artifacts/packages
[2019/03/20 16:10:54][INFO]   Persisting no-op dg to /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/artifacts/obj/MicroBenchmarks/MicroBenchmarks.csproj.nuget.dgspec.json
[2019/03/20 16:10:54][INFO]   Persisting no-op dg to /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/artifacts/obj/BenchmarkDotNet.Extensions/BenchmarkDotNet.Extensions.csproj.nuget.dgspec.json
[2019/03/20 16:19:12][INFO] /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/tools/dotnet/x64/sdk/3.0.100-preview4-010918/NuGet.targets(119,5): error : Unable to load the service index for source https://dotnet.myget.org/F/symreader-converter/api/v3/index.json. [/home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/src/benchmarks/micro/MicroBenchmarks.csproj]
[2019/03/20 16:19:12][INFO] /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/tools/dotnet/x64/sdk/3.0.100-preview4-010918/NuGet.targets(119,5): error :   An error occurred while sending the request. [/home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/src/benchmarks/micro/MicroBenchmarks.csproj]
[2019/03/20 16:19:12][INFO] /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/tools/dotnet/x64/sdk/3.0.100-preview4-010918/NuGet.targets(119,5): error :   Unable to read data from the transport connection: Connection reset by peer. [/home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/src/benchmarks/micro/MicroBenchmarks.csproj]
[2019/03/20 16:19:12][INFO] /home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/tools/dotnet/x64/sdk/3.0.100-preview4-010918/NuGet.targets(119,5): error :   Connection reset by peer [/home/ddit/dotnetbuild/work/cbdb4e6b-e558-4715-ac11-769e2a0162ab/Work/51a9e9bd-019d-491a-b4a8-9990a61b7b8e/Exec/src/benchmarks/micro/MicroBenchmarks.csproj]
[2019/03/20 16:19:12][ERROR] Process exited with status 1
```

That looks like what is probably a transient failure. It is probably a good idea to go through our automation for places like this and add a small retry so we don't lose test runs for simple network trouble.

--- a/scripts/dotnet.py
+++ b/scripts/dotnet.py
@@ -289,7 +289,7 @@ class CSharpProject:
         if runtime_identifier:
             cmdline += ['--runtime', runtime_identifier]
             
-        RunCommand(cmdline, verbose=verbose).run(
+        RunCommand(cmdline, verbose=verbose, retry=1).run(
             self.working_directory)
 
     def build(self,
@@ -706,7 +706,7 @@ def install(
     if versions:
         for version in versions:
             cmdline_args = common_cmdline_args + ['-Version', version]
-            RunCommand(cmdline_args, verbose=verbose).run(
+            RunCommand(cmdline_args, verbose=verbose, retry=1).run(
                 get_repo_root_path()
             )
 
@@ -718,7 +718,7 @@ def install(
     if (not versions) and channels:
         for channel in channels:
             cmdline_args = common_cmdline_args + ['-Channel', channel]
-            RunCommand(cmdline_args, verbose=verbose).run(
+            RunCommand(cmdline_args, verbose=verbose, retry=1).run(
                 get_repo_root_path()
             )
 

--- a/scripts/performance/common.py
+++ b/scripts/performance/common.py
@@ -123,7 +123,8 @@ class RunCommand:
             self,
             cmdline: list,
             success_exit_codes: list = None,
-            verbose: bool = False):
+            verbose: bool = False,
+            retry: int = 0):
         if cmdline is None:
             raise TypeError('Unspecified command line to be executed.')
         if not cmdline:
@@ -131,6 +132,7 @@ class RunCommand:
 
         self.__cmdline = cmdline
         self.__verbose = verbose
+        self.__retry = retry
 
         if success_exit_codes is None:
             self.__success_exit_codes = [0]
@@ -155,8 +157,7 @@ class RunCommand:
         '''Enables/Disables verbosity.'''
         return self.__verbose
 
-    def run(self, working_directory: str = None) -> None:
-        '''Executes specified shell command.'''
+    def __runinternal(self, working_directory: str = None) -> tuple:
         should_pipe = self.verbose
         with push_dir(working_directory):
             quoted_cmdline = '$ '
@@ -180,9 +181,20 @@ class RunCommand:
                                 getLogger().info(line)
 
                     proc.wait()
+                    return (proc.returncode, quoted_cmdline)
 
-                    if proc.returncode not in self.success_exit_codes:
-                        getLogger().error(
-                            "Process exited with status %s", proc.returncode)
-                        raise CalledProcessError(
-                            proc.returncode, quoted_cmdline)
+
+    def run(self, working_directory: str = None) -> None:
+        '''Executes specified shell command.'''
+
+        retrycount = 0
+        (returncode, quoted_cmdline) = self.__runinternal(working_directory)
+        while returncode not in self.success_exit_codes and retrycount < self.__retry:
+            (returncode, _) = self.__runinternal(working_directory)
+            retrycount += 1
+
+        if returncode not in self.success_exit_codes:
+            getLogger().error(
+                "Process exited with status %s", returncode)
+            raise CalledProcessError(
+                returncode, quoted_cmdline)

